kernel/mem_domain.c

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <kernel.h>
+#include <kernel_structs.h>
+#if defined(CONFIG_ARM_CORE_MPU)
+#include <arch/arm/cortex_m/mpu/arm_core_mpu.h>
+#endif
+
+
+static u8_t max_partitions;
+/* Function pointer points to arch/soc specific memory domain configuring API */
+static void (*arch_configure_mem_domain)(struct k_thread *thread);
+
+
+/* return 1 if the given domain is the current thread belongs to, 0 otherwise */
+static inline int is_domain_of_current_thread(struct k_mem_domain *domain)
+{
+	return _current->mem_domain_info.mem_domain == domain ? 1 : 0;
+}
+
+
+int k_mem_domain_init(struct k_mem_domain *domain, const char *name,
+		      u32_t num_parts, struct k_mem_partition *parts[])
+{
+	unsigned int key;
+
+	if (!domain || !name || (num_parts && !parts)) {
+		return -EINVAL;
+	}
+
+	if (num_parts > max_partitions) {
+		return -ENOMEM;
+	}
+
+	key = irq_lock();
+
+	domain->name = name;
+	domain->num_partitions = num_parts;
+	memset(domain->partitions, 0, sizeof(domain->partitions));
+
+	if (num_parts) {
+		u32_t i;
+
+		for (i = 0; i < num_parts; i++) {
+			if (!parts[i]) {
+				irq_unlock(key);
+				return -EINVAL;
+			}
+			domain->partitions[i] = *parts[i];
+		}
+	}
+
+	sys_dlist_init(&domain->mem_domain_q);
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_destroy(struct k_mem_domain *domain)
+{
+	unsigned int key;
+	sys_dnode_t *node, *next_node;
+
+	if (!domain) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	/* Remove threads belong to this domain */
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
+		struct k_thread *thread =
+			CONTAINER_OF(node, struct k_thread, mem_domain_info);
+
+		sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node);
```

Comment:
`sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node); 
thread->mem_domain_info.mem_domain = NULL; `

Can both of this with the 'arch_configure_mem_domain' API form a new API. In case of x86 we need the partition information in order to reset the page tables.